### PR TITLE
Fix cancel builds

### PR DESCRIPTION
**Pull request by ricardoamador**
Round 2 for a fix for cancelling builds on a closed, merged or updated pull request. There were a few issues with the old implementation that I had found in test that this fix addresses:

1. Do not use a Map for the return value when attempting to collect builds. A developer can run multiple builds for a task several times result in bad/stale data being returned in the map for a build, i.e. you are looking for a scheduled build but will get a previously successful build.
2. The old method was specifying the user as cocoon which was not correct.
3. We are not shortcutting the logic on the state of the builds.

I have validated the fix previous to this pull request:
image

But will need to validate one more time before merging this pull request into the prod code.

Validation that we are cancelling builds has completed. I have verified that when we close a pull request, update a pull request that we are cancelling builds and not just stacking new builds. I have also validated that we are starting new requests when we reopen a pull request.

image

image

Hopefully this also reduces our wait on builds and quota since we are not just spawning new builds each time.

*List which issues are fixed by this PR. You must list at least one issue.*
Fixes https://github.com/flutter/flutter/issues/124238

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read the [Flutter Style Guide] _recently_, and have followed its advice.
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: app_dart/lib/src/service/scheduler.dart [hunk 1]
```diff
@@ -279,13 +279,6 @@ class Scheduler {
     String reason = 'Newer commit available',
     List<String>? builderTriggerList,
   }) async {
-    // Always cancel running builds so we don't ever schedule duplicates.
-    log.fine('Attempting to cancel existing presubmit targets for ${pullRequest.number}');
```
Comment (CaseyHillers) [inline app_dart/lib/src/service/scheduler.dart:283]: Is this diff expected? Won't it cause duplicates to be scheduled?

File: app_dart/lib/src/request_handlers/github/webhook_subscription.dart [hunk 2]
```diff
@@ -113,10 +113,16 @@ class GithubWebhookSubscription extends SubscriptionHandler {
         // We'll leave unfinished jobs if it was merged since we care about those
         // results.
         if (!pr.merged!) {
-          await scheduler.cancelPreSubmitTargets(pullRequest: pr, reason: 'Pull request closed');
+          await scheduler.cancelPreSubmitTargets(
```
Comment (CaseyHillers) [inline app_dart/lib/src/request_handlers/github/webhook_subscription.dart:116]: Is there a point to have the if/else logic contain cancelPresubmitTargets?

You can move it above to do it regardless, and if the reason is useful, do `reason: 'Pull request ${pr.merged ? 'merged' : 'closed'}'`

File: app_dart/lib/src/service/luci_build_service.dart [hunk 3]
```diff
@@ -248,20 +200,40 @@ class LuciBuildService {
...
   Future<void> cancelBuilds(github.PullRequest pullRequest, String reason) async {
-    final Map<String?, Build?> builds = await tryBuildsForPullRequest(pullRequest);
-    if (!builds.values.any((Build? build) {
-      return build!.status == Status.scheduled || build.status == Status.started;
-    })) {
+    log.info(
+      'Attempting to cancel builds for pullrequest ${pullRequest.base!.repo!.fullName}/${pullRequest.number}',
+    );
+
+    // final Map<String?, Build?> builds = await tryBuildsForPullRequest(pullRequest);
```
Comment (CaseyHillers) [inline app_dart/lib/src/service/luci_build_service.dart:207]: rm dead code

File: app_dart/lib/src/service/luci_build_service.dart [hunk 4]
```diff
@@ -248,20 +200,40 @@ class LuciBuildService {
...
+    final Iterable<Build> builds = await getTryBuilds(
+      github.RepositorySlug.full(pullRequest.base!.repo!.fullName),
```
Comment (CaseyHillers) [inline app_dart/lib/src/service/luci_build_service.dart:209]:
```suggestion
      pullRequest.base!.repo!.slug(),
```

File: app_dart/lib/src/service/luci_build_service.dart [hunk 5]
```diff
@@ -270,10 +242,14 @@ class LuciBuildService {
     github.PullRequest pullRequest,
     List<Target> targets,
   ) async {
-    final Map<String?, Build?> builds = await tryBuildsForPullRequest(pullRequest);
+    final Iterable<Build> builds = await getTryBuilds(
+      github.RepositorySlug.full(pullRequest.base!.repo!.fullName),
```
Comment (CaseyHillers) [inline app_dart/lib/src/service/luci_build_service.dart:246]:
```suggestion
      pullRequest.base!.repo!.slug(),
```

File: app_dart/lib/src/request_handlers/github/webhook_subscription.dart [hunk 6]
```diff
@@ -138,6 +145,10 @@ class GithubWebhookSubscription extends SubscriptionHandler {
       case 'synchronize':
         // This indicates the PR has new commits. We need to cancel old jobs
         // and schedule new ones.
+        await scheduler.cancelPreSubmitTargets(
```
Comment (CaseyHillers) [inline app_dart/lib/src/request_handlers/github/webhook_subscription.dart:148]: Can you add a test to verify this?

File: app_dart/lib/src/request_handlers/github/webhook_subscription.dart (hunk 2 above)
Comment (ricardoamador) [reply app_dart/lib/src/request_handlers/github/webhook_subscription.dart:116]: No. I will update the code.

File: app_dart/lib/src/service/luci_build_service.dart (hunk 4 above)
Comment (ricardoamador) [reply app_dart/lib/src/service/luci_build_service.dart:209]: Done. I will push the change.

File: app_dart/lib/src/service/luci_build_service.dart (hunk 5 above)
Comment (ricardoamador) [reply app_dart/lib/src/service/luci_build_service.dart:246]: Done. I will push the change.

File: app_dart/lib/src/service/scheduler.dart (hunk 1 above)
Comment (ricardoamador) [reply app_dart/lib/src/service/scheduler.dart:283]: Hmm... Looking at it from the point of view of subscription I don't think it as the cancel builds call was pull out so the method is really only triggering try builds and not cancelling then triggering try builds. I suppose I can add the cancel calls to the other spots where trigger try builds is called. That way the method agreement is only triggering presubmit builds.

Comment (ricardoamador) [reply app_dart/lib/src/service/scheduler.dart:283]: Ah I will add it back. The calls come from multiple entry points into the service expecting this behavior. I can refactor it later.

File: app_dart/lib/src/request_handlers/github/webhook_subscription.dart (hunk 6 above)
Comment (ricardoamador) [reply app_dart/lib/src/request_handlers/github/webhook_subscription.dart]: Reverted this call back to the triggerPresubmitBuilds method.

File: app_dart/lib/src/request_handlers/github/webhook_subscription.dart [hunk 7]
```diff
@@ -171,8 +172,16 @@ class GithubWebhookSubscription extends SubscriptionHandler {
...
-        await gitHubClient.issues.createComment(slug, pr.number!, body);
+      if (!await _alreadyCommented(
+        gitHubClient,
+        pr,
+        body,
+      )) {
+        await gitHubClient.issues.createComment(
+          slug,
+          pr.number!,
+          body,
```
Comment (CaseyHillers) [inline app_dart/lib/src/request_handlers/github/webhook_subscription.dart:183]: Generally, if a method is all positional args, you avoid the trailing comma to make it easier to read

File: app_dart/lib/src/service/luci_build_service.dart [hunk 8]
```diff
@@ -256,20 +208,39 @@ class LuciBuildService {
...
     }
+
     final List<Request> requests = <Request>[];
-    for (Build? build in builds.values) {
-      requests.add(
-        Request(
-          cancelBuild: CancelBuildRequest(id: build!.id, summaryMarkdown: reason),
-        ),
-      );
+    // for (Build? build in builds.values) {
```
Comment (CaseyHillers) [inline app_dart/lib/src/service/luci_build_service.dart:228]: rm dead code?

**Review by CaseyHillers — APPROVED** [review]
LGTM

A lot of this diff is mostly adding trailing commas to positional functions, which I would go against. We only do trailing commas on named arguments

File: app_dart/lib/src/request_handlers/github/webhook_subscription.dart (hunk 7 above)
Comment (ricardoamador) [reply app_dart/lib/src/request_handlers/github/webhook_subscription.dart]: Fixed.

File: app_dart/lib/src/service/luci_build_service.dart (hunk 8 above)
Comment (ricardoamador) [reply app_dart/lib/src/service/luci_build_service.dart:228]: Yes. Thank you. removed.

**Pull request merged**